uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+  let eth_usdc_price_result = get_get_last_snapshot_value_in_eth_usdc_price (targets . get (3usize) . unwrap () . clone ()) . await ;
+  
+  let fees_24h_usd = pool_fees_result.clone().unwrap().result[0].fees_24h_usd;
+  let t0_decimals = pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t0_decimals;
+  let t1_decimals =  pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t1_decimals;
+
+  let eth_price_usd  = eth_usdc_price_result.clone().unwrap().result.t1_price_usd;
+
+  let address = v3pool_result.clone().unwrap().result.address;
+  let current_tick_liquidity_str = v3pool_result.clone().unwrap().result.liquidity;
+  let current_tick_liquidity = i128::from_str_radix(current_tick_liquidity_str.trim_start_matches("0x"), 16).unwrap();
+  let sqrt_ratio_x96_str = v3pool_result.clone().unwrap().result.sqrt_ratio_x96;
+  let sqrt_ratio_x96 = u128::from_str_radix(sqrt_ratio_x96_str.trim_start_matches("0x"), 16).unwrap();
+  let tick_current = v3pool_result.clone().unwrap().result.tick_current;
+  let tick_spacing = v3pool_result.clone().unwrap().result.tick_spacing;
+  let token0 = v3pool_result.clone().unwrap().result.token0;
+
+  let tick_cumul_28x6h = tc28x6_result.unwrap().0;
+
+  let current_sqrt_price = sqrt_ratio_x96 as f32 / f32::powf(2.0,96.0);
+  let mut current_price = current_sqrt_price * current_sqrt_price;
+
+  let mut compressed: i32 = tick_current / tick_spacing;
+  if tick_current <0 && tick_current % tick_spacing != 0 {
+    compressed -= 1;
+  }
+  let floor = compressed * tick_spacing;
+  let mut min_tick = floor;
+  let mut max_tick = floor + tick_spacing;
+  for i in (1..29).rev() {
+    let new_tick = ((tick_cumul_28x6h[i] - tick_cumul_28x6h[i-1]) / 21600) as i32;
+    if new_tick < min_tick {
+      min_tick = new_tick;
+    } else if new_tick > max_tick {
+      max_tick = new_tick;
+    }
+  }
+  
+  let mut range_top = f32::powf(1.0001, max_tick as f32);
+  let mut range_bottom = f32::powf(1.0001, min_tick as f32);
+  let amount0 = current_tick_liquidity as f32
+    * (range_top.sqrt() - current_sqrt_price as f32)
+    / (current_sqrt_price as f32 * range_top.sqrt())
+    / f32::powf(10.0, t0_decimals as f32);
+  let amount1 = current_tick_liquidity as f32
+    * (current_sqrt_price as f32 - range_bottom.sqrt())
+    / f32::powf(10.0, t1_decimals as f32);
+  let est_tvl_in_range = amount0 * current_price + amount1;
+
+  if token0 == "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2" {
+    current_price = 1.0 / current_price;
+    let est_tvl_in_range = amount0 + amount1 * current_price;
+    let new_range_top = 1.0 / range_bottom;
+    let new_range_bottom = 1.0 / range_top;
+    range_top = new_range_top;
+    range_bottom = new_range_bottom;
+  }
+  
+  let fees_24h_eth = fees_24h_usd / eth_price_usd;
+  let edr = fees_24h_eth / est_tvl_in_range as f32;
+  let edpr = edr * 100.0;
+
+  let result = LensValue {
```

Comment:
The logic that determines token0 and inverts some values in the preceding process is the reason for using mutable variables, but mutable variables reduce readability (they can change values in all valid scopes, making code reading difficult). (Code reading is difficult because the value can change in all scopes in which the variable is valid.)

I think adding it here would eliminate the need for mutable variables.
I think it goes something like this.
```rust
let result = if token0 == "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2" {
  LensValue {
    address: address.to_string(),
    current_price: 1.0 / current_price,
    range_top: 1.0 / range_bottom,
    range_bottom: 1.0 / range_top,
    edpr
  }
} else {
  LensValue {
    address: address.to_string(),
    current_price,
    range_top,
    range_bottom,
    edpr
  }  
}
```

Reply:
Yes, that looks better. However when tried out in the [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=45056dfb54dab7e174738232579cdaf1) it returns the previous incorrect result for `edpr`. I suspect this is because `est_tvl_in_range` also needs to be adjusted. I imagine there's an easy fix for this but I'll leave it alone for now.